benchmark/UnsortedJoinBenchmark.cpp

```diff
@@ -0,0 +1,96 @@
+#include "util/Forward.h"
+#include "engine/idTable/IdTable.h"
+#include "../benchmark/Benchmark.h"
+
+// Local helper function.
+#include "../benchmark/util/IdTableHelperFunction.h"
+
+// Non-local helper function.
+#include "../test/util/IdTableHelpers.h"
+#include "../test/util/JoinHelpers.h"
+#include "../test/IndexTestHelpers.h"
+
+// Benchmarks for unsorted tables, with and without overlapping values in
+// IdTables. Done with normal join and hash join.
+void BM_UnsortedIdTable(BenchmarkRecords* records) {
+  // For easier changing of the IdTables size.
+  // Important: Apparently Join::join has an inbuild timeout checker, which
+  // keeps getting triggerd by tables with a size between 500-600, or higher.
+  // This leads to a segmentation fault, because it keeps trying to throw an
+  // exception about the time out, but that exception tries to include
+  // information about the query execution tree using a shared pointer. And, as
+  // far as I can tell, that pointer was never actually initialized here, so
+  // it points into memory, which we are not allowed to access. Which leads
+  // to the seg fault.
+  const size_t NUMBER_ROWS = 500;
+  const size_t NUMBER_COLUMNS = NUMBER_ROWS; 
+
+  Join J{Join::InvalidOnlyForTestingJoinTag{}, ad_utility::testing::getQec()};
+  DISABLE_WARNINGS_CLANG_13
+  auto hashJoinLambda = [&J]<int A, int B, int C>(auto&&... args) {
+    ENABLE_WARNINGS_CLANG_13
+    return J.hashJoin(AD_FWD(args)...);
+  };
+  DISABLE_WARNINGS_CLANG_13
+  auto joinLambda = [&J]<int A, int B, int C>(auto&&... args) {
+    ENABLE_WARNINGS_CLANG_13
+    return J.join<A, B, C>(AD_FWD(args)...);
+  };
+
+  // Tables, that have overlapping values in their join columns.
+  IdTableAndJoinColumn a{
+    createRandomlyFilledIdTable(NUMBER_ROWS, NUMBER_COLUMNS, 0, 0, 10), 0};
+  IdTableAndJoinColumn b{
+    createRandomlyFilledIdTable(NUMBER_ROWS, NUMBER_COLUMNS, 0, 5, 15), 0};
+  
+  // Because overlap is not yet guaranteed, we put some in.
+  for (size_t i = 0; i*20 < NUMBER_ROWS; i++) {
+    a.idTable(i*10, 0) = I(10);
+    b.idTable(i*20, 0) = I(10);
+  }
+
+  // We need to order the IdTables, before using the normal join function on
+  // them. This is the sort function for that.
+  auto sortFunction = [](const auto& row1, const auto& row2) {
+    return row1[0] < row2[0];
+  };
+
+  // Lambda wrapper for the functions, that I measure.
+  auto joinLambdaWrapper = [&]() {
+        // Sorting the tables after the join column.
+        std::sort(a.idTable.begin(), a.idTable.end(), sortFunction);
+        std::sort(b.idTable.begin(), b.idTable.end(), sortFunction);
+
+        useJoinFunctionOnIdTables(a, b, joinLambda);
+  };
+  auto hashJoinLambdaWrapper = [&]() {
+        useJoinFunctionOnIdTables(a, b, hashJoinLambda);
+  };
+
+  records->addSingleMeasurment("Hashed join with overlapping IdTables", hashJoinLambdaWrapper);
+  
+  records->addSingleMeasurment("Normal join with overlapping IdTables", joinLambdaWrapper);
+
```

Comment:
I think you almost always want to run the same tests on sorted and on unsorted tables, why not merge the two functions, create the tables only once, and `random_shuffle` them in between the two benchmarks, like you do it in the tests?

Reply:
I mainly split it in two files, because one should keep benchmarks seperate, so that you don't have to wait for multiple benchmarks to run through, if you only want to know the result of one.
I'm honestly not sure, if that is less, or more important than having no code duplication.

I'm gonna make a new separat benchmark, that implements your suggestion and we can then decide, which one we should keep.

Reply:
Wrote the new benchmark in [409c8ae](https://github.com/ad-freiburg/qlever/pull/860/commits/409c8ae02fc54ba5cae35658873b180ad7a1857d).